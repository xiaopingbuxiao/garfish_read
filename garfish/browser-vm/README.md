# `@garfish/browser-vm`

[![NPM version](https://img.shields.io/npm/v/@garfish/browser-vm.svg?style=flat-square)](https://www.npmjs.com/package/@garfish/browser-vm)

## Usage

```js
import vmSandbox from '@garfish/browser-vm';

let nameMap = { a: 'chen' };

const sandbox = new vmSandbox({
  namespace: 'app',
  el: () => document.body,
  modules: [
    () => ({
      override: { nameMap },
      recover() {
        nameMap = { a: 'chen' };
      },
    }),
  ],
});

sandbox.execScript(`
  window.x = 1;
  console.log(window.nameMap); // { a: 'chen' }
  window.nameMap.a = 'tao';
`);

console.log(sandbox.global.x); // 1
console.log(sandbox.global.nameMap); // { a: 'tao' }

// Clear effect
sandbox.clearEffects();
console.log(nameMap); // { a: 'chen' }

// If clear all effects
sandbox.reset();
```

## Hooks

```js
const sandbox = new vmSandbox({
  namespace: 'app',
  el: () => document.body,
});

sandbox.hooks.usePlugin({
  stared(fakeWindow) {},
  closed() {},
  beforeClearEffect() {},
  afterClearEffect() {},
  beforeInvoke(url, env, options) {},
  afterInvoke(url, env, options) {},
  invokeError(err, url, env, options) {},
  appendNode(parentNode, oldNode, convertedNode, tagName) {},
});
```
