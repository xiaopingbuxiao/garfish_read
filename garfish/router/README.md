# `@garfish/router`

[![NPM version](https://img.shields.io/npm/v/@garfish/router.svg?style=flat-square)](https://www.npmjs.com/package/@garfish/router)

## Usage

```js
import router from '@garfish/router';

// TODO: DEMONSTRATE API
```
