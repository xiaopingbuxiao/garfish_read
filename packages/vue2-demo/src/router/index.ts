import type { RouteConfig } from "vue-router";
import HomeView from "../views/HomeView.vue";

export const routes: Array<RouteConfig> = [
  {
    path: "/",
    name: "home",
    component: HomeView,
  },
  {
    path: "/about",
    name: "about",
    // route level code-splitting
    // this generates a separate chunk (about.[hash].js) for this route
    // which is lazy-loaded when the route is visited.
    component: () =>
      import(/* webpackChunkName: "about" */ "../views/AboutView.vue"),
  },
];
