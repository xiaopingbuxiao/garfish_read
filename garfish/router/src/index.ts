import type { interfaces } from '@garfish/core';
import { createKey, routerLog } from '@garfish/utils';
import { RouterConfig } from './config';
import router, {
  initRedirect,
  RouterInterface,
  listenRouterAndReDirect,
} from './context';

declare module '@garfish/core' {
  export default interface Garfish {
    router: RouterInterface;
    apps: Record<string, interfaces.App>;
  }

  export namespace interfaces {
    export interface Config {
      autoRefreshApp?: boolean;
      onNotMatchRouter?: (path: string) => Promise<void> | void;
    }

    export interface AppInfo {
      activeWhen?: string | ((path: string) => boolean); // 手动加载，可不填写路由
      active?: (appInfo: AppInfo, rootPath: string) => void;
      deactive?: (appInfo: AppInfo, rootPath: string) => void;
      rootPath?: string;
      basename?: string;
    }
  }
}

export type { RouterInterface } from './context';

interface Options {
  autoRefreshApp?: boolean;
  onNotMatchRouter?: (path: string) => Promise<void> | void;
}

export function GarfishRouter(_args?: Options) {
  return function (Garfish: interfaces.Garfish): interfaces.Plugin {
    Garfish.apps = {};
    Garfish.router = router;

    return {
      name: 'router',
      version: __VERSION__,

      bootstrap(options: interfaces.Options) {
        let activeApp: null | string = null;
        const unmounts: Record<string, Function> = {};
        const { basename } = options;
        const { autoRefreshApp = true, onNotMatchRouter = () => null } =
          Garfish.options;

        async function active(
          appInfo: interfaces.AppInfo,
          rootPath: string = '/',
        ) {
          routerLog(`${appInfo.name} active`, {
            appInfo,
            rootPath,
            listening: RouterConfig.listening,
          });

          // In the listening state, trigger the rendering of the application
          if (!RouterConfig.listening) return;

          const { name, active, cache = true } = appInfo;
          if (active) return active(appInfo, rootPath);
          appInfo.rootPath = rootPath;

          const currentApp = (activeApp = createKey());
          const app = await Garfish.loadApp(appInfo.name, {
            cache,
            basename: rootPath,
            entry: appInfo.entry,
            domGetter: appInfo.domGetter,
          });

          if (app) {
            app.appInfo.basename = rootPath;

            const call = async (app: interfaces.App, isRender: boolean) => {
              if (!app) return;
              const isDes = cache && app.mounted;
              if (isRender) {
                return await app[isDes ? 'show' : 'mount']();
              } else {
                return app[isDes ? 'hide' : 'unmount']();
              }
            };

            Garfish.apps[name] = app;
            unmounts[name] = () => {
              // Destroy the application during rendering and discard the application instance
              if (app.mounting) {
                delete Garfish.cacheApps[name];
              }
              call(app, false);
            };

            if (currentApp === activeApp) {
              await call(app, true);
            }
          }
        }

        async function deactive(appInfo: interfaces.AppInfo, rootPath: string) {
          routerLog(`${appInfo.name} deactive`, {
            appInfo,
            rootPath,
          });

          activeApp = null;
          const { name, deactive } = appInfo;
          if (deactive) return deactive(appInfo, rootPath);

          const unmount = unmounts[name];
          unmount && unmount();
          delete Garfish.apps[name];

          // Nested scene to remove the current application of nested data
          // To avoid the main application prior to application
          const needToDeleteApps = router.routerConfig.apps.filter((app) => {
            if (appInfo.rootPath === app.basename) return true;
          });
          if (needToDeleteApps.length > 0) {
            needToDeleteApps.forEach((app) => {
              delete Garfish.appInfos[app.name];
              delete Garfish.cacheApps[app.name];
            });
            router.setRouterConfig({
              apps: router.routerConfig.apps.filter((app) => {
                return !needToDeleteApps.some(
                  (needDelete) => app.name === needDelete.name,
                );
              }),
            });
          }
        }

        const apps = Object.values(Garfish.appInfos);

        const appList = apps.filter((app) => {
          if (!app.basename) app.basename = basename;
          return !!app.activeWhen;
        }) as Array<Required<interfaces.AppInfo>>;

        const listenOptions = {
          basename,
          active,
          deactive,
          autoRefreshApp,
          notMatch: onNotMatchRouter,
          apps: appList,
          listening: true,
        };
        routerLog('listenRouterAndReDirect', listenOptions);
        listenRouterAndReDirect(listenOptions);
      },

      registerApp(appInfos) {
        const appList = Object.values(appInfos);
        // @ts-ignore
        router.registerRouter(appList.filter((app) => !!app.activeWhen));
        // After completion of the registration application, trigger application mount
        // Has been running after adding routing to trigger the redirection
        if (!Garfish.running) return;
        routerLog('registerApp initRedirect', appInfos);
        initRedirect();
      },
    };
  };
}
