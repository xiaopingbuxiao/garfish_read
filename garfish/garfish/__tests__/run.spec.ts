import assert from 'assert';
import Garfish, { interfaces } from '@garfish/core';
import { GarfishRouter } from '@garfish/router';
import { GarfishBrowserVm } from '@garfish/browser-vm';
import { mockStaticServer } from '@garfish/test-suite';
import {
  __MockBody__,
  __MockHead__,
  __MockHtml__,
  appContainerId,
} from '@garfish/utils';

const vueAppRootNode = 'vue-app';
const vueAppRootText = 'vue app init page';
const reactAppRootNode = 'react-app';
const reactAppRootText = 'react app init page';

function waitFor(delay = 50) {
  return new Promise<void>((resolve) => {
    setTimeout(() => resolve(), delay);
  });
}

async function vueAppInDocument(container: Element) {
  await waitFor();
  const appContainer = container.querySelectorAll(`[id^=${appContainerId}]`);
  expect(appContainer).toHaveLength(1);
  assert(appContainer[0]);
  assert(appContainer[0].querySelector(`#${vueAppRootNode}`));
  expect(appContainer[0].querySelector(`#${vueAppRootNode}`)?.innerHTML).toBe(
    vueAppRootText,
  );
}

async function reactAppInDocument(container: Element) {
  await waitFor();
  const appContainer = container.querySelectorAll(`[id^=${appContainerId}]`);
  expect(appContainer).toHaveLength(1);
  assert(appContainer[0].querySelector(`#${reactAppRootNode}`));
  expect(appContainer[0].querySelector(`#${reactAppRootNode}`)?.innerHTML).toBe(
    reactAppRootText,
  );
}

async function noRenderApp(container: Element) {
  await waitFor();
  const appContainer = container.querySelectorAll(`[id^=${appContainerId}]`);
  expect(appContainer).toHaveLength(0);
}

// 业务自定义garfish loader
export function GarfishLoader() {
  return function (Garfish: interfaces.Garfish): interfaces.Plugin {
    return {
      name: 'garfish-loader',
      bootstrap(options: interfaces.Options = {}) {
        Garfish.loader.usePlugin(
          Object.assign({ name: 'garfish-loader' }, options.loader),
        );
      },
    };
  };
}

const mockBeforeLoad = jest.fn();
const mockAfterLoad = jest.fn();
const mockBeforeMount = jest.fn();
const mockAfterMount = jest.fn();
const mockBeforeUnmount = jest.fn();
const mockAfterUnmount = jest.fn();
const mockBeforeEval = jest.fn();
const mockAfterEval = jest.fn();

const loaderFetchReturn = {
  code: 'function() {}',
  type: 'aplication/javascript',
  size: 13,
};

describe('Core: run methods', () => {
  let GarfishInstance: Garfish;

  mockStaticServer({
    baseDir: __dirname,
  });

  beforeEach(() => {
    GarfishInstance = new Garfish({});

    GarfishInstance.run({
      basename: '/',
      domGetter: '#container',
      plugins: [GarfishRouter(), GarfishBrowserVm(), GarfishLoader()],
      apps: [
        {
          name: 'vue-app',
          activeWhen: '/vue-app',
          entry: './resources/vueApp.html',
        },
        {
          name: 'react-app',
          activeWhen: '/react-app',
          entry: './resources/reactApp.html',
        },
      ],
      beforeLoad: mockBeforeLoad,
      afterLoad: mockAfterLoad,
      beforeEval: mockBeforeEval,
      afterEval: mockAfterEval,
      beforeMount: mockBeforeMount,
      afterMount: mockAfterMount,
      beforeUnmount: mockBeforeUnmount,
      afterUnmount: mockAfterUnmount,
      loader: {
        fetch: async (url) => {
          if (url === 'http://jest') {
            return new Response(loaderFetchReturn.code, {
              status: 200,
              headers: {
                'Content-Type': loaderFetchReturn.type,
              },
            });
          }
        },
      },
    });
  });

  it('auto render and destroy vue app', async () => {
    const container = document.createElement('div');
    container.setAttribute('id', 'container');
    document.body.appendChild(container);

    GarfishInstance.router.push({ path: '/vue-app' });
    await vueAppInDocument(container);
    await waitFor(1000);
    expect(mockBeforeLoad.mock.calls[0][0].name).toBe('vue-app');
    expect(mockAfterLoad.mock.calls[0][0].name).toBe('vue-app');
    expect(mockBeforeMount.mock.calls[0][0].name).toBe('vue-app');
    expect(mockAfterMount.mock.calls[0][0].name).toBe('vue-app');
    expect(mockBeforeEval.mock.calls[0][0].name).toBe('vue-app');
    expect(
      mockBeforeEval.mock.calls.map((callArgs) => {
        return callArgs[3];
      }),
    ).toMatchSnapshot();

    GarfishInstance.router.push({ path: '/react-app' });
    await reactAppInDocument(container);
    expect(mockBeforeMount.mock.calls[0][0].name).toBe('vue-app');
    expect(mockAfterMount.mock.calls[0][0].name).toBe('vue-app');

    expect(mockBeforeLoad.mock.calls[1][0].name).toBe('react-app');
    expect(mockAfterLoad.mock.calls[1][0].name).toBe('react-app');
    expect(mockBeforeMount.mock.calls[1][0].name).toBe('react-app');
    expect(mockAfterMount.mock.calls[1][0].name).toBe('react-app');

    GarfishInstance.router.push({ path: '/' });
    await noRenderApp(container);
    expect(mockBeforeMount.mock.calls[1][0].name).toBe('react-app');
    expect(mockAfterMount.mock.calls[1][0].name).toBe('react-app');

    document.body.removeChild(container);
  });

  it('close router listening', async () => {
    const container = document.createElement('div');
    container.setAttribute('id', 'container');
    document.body.appendChild(container);
    await noRenderApp(container);

    GarfishInstance.router.push({ path: '/react-app' });
    await reactAppInDocument(container);

    GarfishInstance.router.setRouterConfig({ listening: false });
    GarfishInstance.router.push({ path: '/vue-app' });
    await noRenderApp(container);
    GarfishInstance.router.push({ path: '/react-app' });
    await noRenderApp(container);
    GarfishInstance.router.push({ path: '/' });
    await noRenderApp(container);

    GarfishInstance.router.setRouterConfig({ listening: true });
    GarfishInstance.router.push({ path: '/vue-app' });
    await vueAppInDocument(container);
  });

  it('custom fetch', async () => {
    const fetchResult = await GarfishInstance.loader.load({
      scope: 'jest-loader-fetch',
      url: 'http://jest',
    });

    expect(fetchResult.code).toBe(loaderFetchReturn.code);
  });
});
