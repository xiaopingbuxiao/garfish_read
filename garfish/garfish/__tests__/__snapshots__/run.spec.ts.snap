// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core: run methods auto render and destroy vue app 1`] = `
Array [
  "http://localhost/resources/vueIndex.js",
  "./resources/vueApp.html",
  undefined,
  "http://localhost/resources/dynamic.js",
]
`;
